pyomo/contrib/gdpopt/util.py

```diff
@@ -253,41 +153,126 @@ def copy_var_list_values(from_list, to_list, config,
     """
+    if ignore_integrality:
+        deprecation_warning("The 'ignore_integrality' argument no longer "
+                            "has any functionality.", version="TBD")
+
     if len(from_list) != len(to_list):
         raise ValueError('The lengths of from_list and to_list do not match.')
+
     for v_from, v_to in zip(from_list, to_list):
         if skip_stale and v_from.stale:
             continue  # Skip stale variable values.
         if skip_fixed and v_to.is_fixed():
             continue  # Skip fixed variables.
-        try:
-            # We don't want to trigger the reset of the global stale
-            # indicator, so we will set this variable to be "stale",
-            # knowing that set_value will switch it back to "not
-            # stale"
-            v_to.stale = True
-            # NOTE: PEP 2180 changes the var behavior so that domain /
-            # bounds violations no longer generate exceptions (and
-            # instead log warnings).  This means that the following will
-            # always succeed and the ValueError should never be raised.
-            v_to.set_value(value(v_from, exception=False), skip_validation=True)
-        except ValueError as err:
-            err_msg = getattr(err, 'message', str(err))
-            var_val = value(v_from)
-            rounded_val = int(round(var_val))
-            # Check to see if this is just a tolerance issue
-            if ignore_integrality and v_to.is_integer():
-                v_to.set_value(var_val, skip_validation=True)
-            elif v_to.is_integer() and (fabs(var_val - rounded_val) <=
-                                        config.integer_tolerance):
-                v_to.set_value(rounded_val, skip_validation=True)
-            elif abs(var_val) <= config.zero_tolerance and 0 in v_to.domain:
-                v_to.set_value(0, skip_validation=True)
-            else:
-                config.logger.error(
-                    'Unknown validation domain error setting variable %s', (v_to.name,))
-                raise
+        v_to.set_value(value(v_from, exception=False), skip_validation=True)
+
+def fix_discrete_var(var, val, config):
+    """Fixes the discrete variable var to val, rounding to the nearest integer
+    or not, depending on if rounding is specifed in config and what the integer
+    tolerance is."""
+    if val is None:
+        return
+    if var.is_continuous():
+        var.set_value(val, skip_validation=True)
+    elif (fabs(val - round(val)) > config.integer_tolerance):
+        raise ValueError(
+            "Integer variable '%s' cannot be fixed to value %s because it "
+            "is not within the specified integer tolerance of %s." %
+            (var.name, val, config.integer_tolerance))
+    else:
+        # variable is integer and within tolerance
+        if config.round_discrete_vars:
+            var.fix(int(round(val)))
+        else:
+            var.fix(val, skip_validation=True)
+
+@contextmanager
```

Comment:
This is a huge context manager.  We should make it an explicit class with `__enter__` / `__exit__` methods to make this more readable.